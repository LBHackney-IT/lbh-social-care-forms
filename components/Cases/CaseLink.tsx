import { mapFormIdToFormDefinition } from 'data/flexibleForms/mapFormIdsToFormDefinition';
import Link from 'next/link';
import { CaseFormData } from 'types';
import { generateInternalLink } from 'utils/urls';
interface Props {
  recordId: string;
  externalUrl?: string;
  caseFormData: CaseFormData;
  formName: string;
  personId: number;
  title?: string;
  formType?: string;
  isImported: boolean;
  deleted: boolean;
}

const CaseLink = ({
  formName,
  externalUrl,
  caseFormData,
  recordId,
  personId,
  title,
  formType,
  isImported,
  deleted,
}: Props): React.ReactElement | null => {
  if (formType === 'flexible-form') {
    const name = mapFormIdToFormDefinition[formName]?.displayName;

    return (
      <Link href={`/people/${personId}/submissions/${recordId}`}>
        <a className="lbh-link">
          {name || 'View'}
          {title ? ` - ${title}` : ''}
        </a>
      </Link>
    );
  }

  if (externalUrl) {
    return (
      <a
        href={externalUrl}
        target="_blank"
        rel="noreferrer noopener"
        className="lbh-link"
      >
        View
      </a>
    );
  }
  const internalLink = generateInternalLink({
    recordId,
    caseFormData,
    personId,
    formName,
    isImported,
    deleted,
  });
  return internalLink ? (
    <Link href={internalLink}>
      <a className="lbh-link">View</a>
    </Link>
  ) : null;
};

export default CaseLink;
