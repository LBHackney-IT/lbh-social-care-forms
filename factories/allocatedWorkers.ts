import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
