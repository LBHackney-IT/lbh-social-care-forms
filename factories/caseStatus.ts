import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
