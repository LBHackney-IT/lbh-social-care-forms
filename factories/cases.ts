import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
