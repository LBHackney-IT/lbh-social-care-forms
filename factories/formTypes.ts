import { Factory } from 'fishery';
import { FormValue, FormStatusValue } from 'types';

export const mockedFormValueFactory = Factory.define<FormValue>(() => ({
  name: 'test',
  description: 'description',
  options: [mockedFormStatusValueFactory.build()],
}));

export const mockedFormStatusValueFactory = Factory.define<FormStatusValue>(
  () => ({
    name: 'N1',
    description: 'Noo 1',
  })
);

export const mockedFormValue = [mockedFormValueFactory.build()];
