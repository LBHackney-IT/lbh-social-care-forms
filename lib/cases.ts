import axios from 'axios';

import type { Case, CaseData, HistoricCaseData, User } from 'types';
import { getAuditingParams } from '../utils/auditing';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headersWithKey = {
  'x-api-key': AWS_KEY,
};

const regex = /"_id.*ObjectId\(\S*\),./;

export const sanitiseCaseFormData = (caseFormData: string): string =>
  caseFormData && JSON.parse(caseFormData.replace(regex, ''));

export const getCases = async (
  params: Record<string, unknown>
): Promise<CaseData> => {
  const { data } = await axios.get(`${ENDPOINT_API}/cases`, {
    headers: headersWithKey,
    params: params,
  });
  return {
    ...data,
    cases: data.cases?.map((c: Record<string, string>) => ({
      ...c,
      caseFormData: sanitiseCaseFormData(c.caseFormData),
    })),
  };
};

export const getCasesByResident = (
  mosaic_id: number,
  params: Record<string, unknown>
): Promise<CaseData> => getCases({ mosaic_id, ...params });

export const getCase = async (
  case_id: string,
  params: Record<string, unknown>,
  user: User
): Promise<Case | undefined> => {
  const { data } = await axios.get(`${ENDPOINT_API}/cases/${case_id}`, {
    headers: headersWithKey,
    params: {
      ...params,
      ...getAuditingParams(user),
    },
  });
  return (
    data && { ...data, caseFormData: sanitiseCaseFormData(data.caseFormData) }
  );
};

export const getHistoricNote = async (
  case_id: string,
  params: Record<string, unknown>
): Promise<HistoricCaseData | undefined> => {
  const { data } = await axios.get(`${ENDPOINT_API}/casenotes/${case_id}`, {
    headers: headersWithKey,
    params,
  });
  return data;
};

export const getHistoricVisit = async (
  case_id: string,
  params: Record<string, unknown>
): Promise<HistoricCaseData | undefined> => {
  const { data } = await axios.get(`${ENDPOINT_API}/visits/${case_id}`, {
    headers: headersWithKey,
    params,
  });
  return data;
};

export const addCase = async (
  formData: Record<string, unknown>
): Promise<{ ref: string }> => {
  const { data } = await axios.post(`${ENDPOINT_API}/cases`, formData, {
    headers: { 'Content-Type': 'application/json', 'x-api-key': AWS_KEY },
  });
  return data;
};
