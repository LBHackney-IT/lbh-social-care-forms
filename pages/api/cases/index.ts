import { StatusCodes } from 'http-status-codes';

import { getCases, addCase } from 'lib/cases';
import { middleware as csrfMiddleware } from 'lib/csrfToken';
import { handleAxiosError } from 'lib/errorHandler';

import { AxiosError } from 'axios';
import { apiHandler, AuthenticatedNextApiHandler } from 'lib/apiHandler';

const endpoint: AuthenticatedNextApiHandler = async (req, res) => {
  switch (req.method) {
    case 'GET':
      try {
        const data = await getCases({
          ...req.query,
          context_flag: req.user.permissionFlag,
        });
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Cases get error:',
          (error as AxiosError)?.response?.data
        );
        res = handleAxiosError(res, error as AxiosError, 'Cases');
      }
      break;

    case 'POST':
      try {
        const data = await addCase(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Case post error:',
          (error as AxiosError)?.response?.data
        );
        res = handleAxiosError(res, error as AxiosError, 'Cases');
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
