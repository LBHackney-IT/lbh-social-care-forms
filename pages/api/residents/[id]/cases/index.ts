import { StatusCodes } from 'http-status-codes';

import { getCasesByResident } from 'lib/cases';
import { middleware as csrfMiddleware } from 'lib/csrfToken';

import { AxiosError } from 'axios';
import {
  apiHandler,
  AuthenticatedNextApiHandler,
  handleAxiosError,
} from 'lib/apiHandler';

const endpoint: AuthenticatedNextApiHandler = async (req, res) => {
  const { id, ...params } = req.query;
  switch (req.method) {
    case 'GET':
      try {
        const data = await getCasesByResident(parseInt(id as string, 10), {
          ...params,
          context_flag: req.user.permissionFlag,
        });
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Cases get error:',
          (error as AxiosError)?.response?.data
        );
        res = handleAxiosError(res, error as AxiosError, 'Cases');
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
