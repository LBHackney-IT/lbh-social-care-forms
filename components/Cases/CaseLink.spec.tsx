import { render } from '@testing-library/react';
import { mockedDeallocationNote, mockedWarningNoteCase } from 'factories/cases';

import CaseLink from './CaseLink';

describe('CaseLink component', () => {
  it('should render properly - with externalUrl', () => {
    const props = {
      recordId: '123',
      externalUrl: 'https://foo.bar',
      caseFormData: mockedDeallocationNote.caseFormData,
      personId: 1,
      formName: 'foo',
    };
    const { asFragment } = render(<CaseLink {...props} />);
    expect(asFragment()).toMatchInlineSnapshot(`
      <DocumentFragment>
        <a
          class="lbh-link"
          href="/people/1/submissions/123"
        >
          Sandbox form
        </a>
      </DocumentFragment>
    `);
  });

  it('should render properly - with handled link', () => {
    const props = {
      recordId: '123',
      caseFormData: mockedDeallocationNote.caseFormData,
      personId: 1,
      formName: 'foo',
    };
    const { asFragment } = render(<CaseLink {...props} />);
    expect(asFragment()).toMatchInlineSnapshot(`
      <DocumentFragment>
        <a
          class="lbh-link"
          href="/people/1/submissions/123"
        >
          Sandbox form
        </a>
      </DocumentFragment>
    `);
  });

  it('should render properly - without a link', () => {
    const props = {
      recordId: '123',
      caseFormData: {
        ...mockedDeallocationNote.caseFormData,
        form_name_overall: '',
      },
      personId: 1,
      formName: 'foo',
    };
    const { asFragment } = render(<CaseLink {...props} />);
    expect(asFragment()).toMatchInlineSnapshot(`
      <DocumentFragment>
        <a
          class="lbh-link"
          href="/people/1/submissions/123"
        >
          Sandbox form
        </a>
      </DocumentFragment>
    `);
  });

  it('should render properly - with handled link for viewing warning notes', () => {
    const props = {
      recordId: '123',
      caseFormData: {
        ...mockedWarningNoteCase.caseFormData,
        form_name: 'Warning Note Created',
      },
      personId: 1,
      formName: 'foo',
    };
    const { asFragment } = render(<CaseLink {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
