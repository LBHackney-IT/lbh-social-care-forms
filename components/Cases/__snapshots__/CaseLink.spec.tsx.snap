// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseLink component should render properly - with handled link for viewing warning notes 1`] = `
<DocumentFragment>
  <a
    class="lbh-link"
    href="/people/123/warning-notes/456/view/note-created"
  >
    View
  </a>
</DocumentFragment>
`;
